[dynamo][guards-cpp-refactor] DictGuardManager fast path using dict tags

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120726
* #120739
* #120730
* #120673
* __->__ #120725
* #120593

Found it messy to use DICT_VERSION guard from Python side with DictGuardManager, so just merging the DICT_VERSION into DictGuardManager.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

anijain2305 committed:
[dynamo][guards-cpp-refactor] DictGuardManager fast path using dict tags

[ghstack-poisoned]

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -1899,13 +1871,20 @@ class DictGuardManager : public GuardManager {
+      return true;
+    } else {
+      // update the tag and then fallback to the slow path.
+      _tag = tag;
```
If the guards below fail, wouldn't it be incorrect to update the tag?

anijain2305 replied:
Yes, good point. We can call this guard with two different unchanging dicts, and this will cause failure every time.

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -1899,13 +1871,20 @@ class DictGuardManager : public GuardManager {
+    int64_t tag = _get_dict_version(obj);
+    if (_tag == tag) {
+      // Fast path for the common case.
+      return true;
```
Is it safe to skip all the guards?

What if you have:
```
x = {"foo": []}
...
# this won't increment the version counter:
x["foo"].append(...)
```

anijain2305 replied:
I don't think there is a good solution here. Anything I come up with is either incomplete or slow/messy to write.

jansel replied:
You need an allowlist of sub-guards that are safe to skip.
1) ID_MATCH
2) TENSOR_MATCH
3) TYPE_MATCH
4) ...

Maybe via a `virtual bool can_skip_with_dict_version() const;`

anijain2305 committed:
Update on "[dynamo][guards-cpp-refactor] DictGuardManager fast path using dict tags"


Found it messy to use DICT_VERSION guard from Python side with DictGuardManager, so just merging the DICT_VERSION into DictGuardManager.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]
